src/serverenvironment.cpp

```diff
@@ -1637,16 +1637,16 @@ u32 ServerEnvironment::addParticleSpawner(float exptime)
 	// Timers with lifetime 0 do not expire
 	float time = exptime > 0.f ? exptime : PARTICLE_SPAWNER_NO_EXPIRY;
 
-	u32 id = 0;
-	for (;;) { // look for unused particlespawner id
-		id++;
-		std::unordered_map<u32, float>::iterator f = m_particle_spawners.find(id);
-		if (f == m_particle_spawners.end()) {
-			m_particle_spawners[id] = time;
-			break;
-		}
+	u32 free_id = m_particle_spawners_id_last_used;
+	while (free_id == 0 || m_particle_spawners.find(free_id) != m_particle_spawners.end()) {
+		if (free_id == m_particle_spawners_id_last_used)
+			return 0; // full
```

Comment:
is zero a valid id?
I suppose not, would be good to use `std::optional<u32>` for the return one day

Reply:
I currently do not see any indicator that 0 is a reserved ID - neither on client- nor server-side. However, it will at least be more consistent with the ID assignment from ActiveObjectMgr.